### PR TITLE
feat: add `node_urls` parameter to bzlmod `toolchain` in the `node` extension

### PR DESCRIPTION
**Pull request by redsun82**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* I don't think I have the capacity to also add tests for this.
* I don't think there's currently any bzlmod documentation, so not sure where I would be adding this.

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

`node_urls` can be provided to add nodejs mirrors to the `WORKFLOW` based setup, but the same can't be done when using bzlmod.

Issue Number: N/A

## What is the new behavior?

When using bzlmod, `node_urls` can be provided to the `toolchain` tag class in the `node` extension, behaving just like the `WORKFLOW` counterpart.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: nodejs/extensions.bzl [hunk 1]
```diff
@@ -62,6 +72,10 @@ If set then the version found in the .nvmrc file is used instead of the one spec
 This setting creates a dependency on a c++ toolchain.
 """,
             ),
+            "node_urls": attr.string_list(
+                doc = "Custom list of URL formats to use to download NodeJS, containing {version} and {filename}",
```
Comment (jbedard) [inline nodejs/extensions.bzl:76]: Can you copy the same docs that are in the repo rule version?
```
List of URLs to use to download Node.js.

Each entry is a template for downloading a node distribution.

The `{version}` parameter is substituted with the `node_version` attribute,
and `{filename}` with the matching entry from the `node_repositories` attribute.
```

File: nodejs/extensions.bzl [hunk 2]
```diff
@@ -1,40 +1,50 @@
...
             else:
-                registrations[toolchain.name] = struct(
-                    node_version = toolchain.node_version,
-                    node_version_from_nvmrc = toolchain.node_version_from_nvmrc,
-                    include_headers = toolchain.include_headers,
-                )
+                registrations[toolchain.name] = toolchain
 
-    for k, v in registrations.items():
+    for k, t in registrations.items():
```
Comment (jbedard) [inline nodejs/extensions.bzl:41]: Any reason for this? Revert if not...

File: nodejs/extensions.bzl [hunk 3]
```diff
@@ -1,40 +1,50 @@
...
+    "nodejs_register_toolchains",
+)
 
 def _toolchain_extension(module_ctx):
     registrations = {}
     for mod in module_ctx.modules:
         for toolchain in mod.tags.toolchain:
             if toolchain.name != DEFAULT_NODE_REPOSITORY and not mod.is_root:
                 fail("Only the root module may provide a name for the node toolchain.")
-
```
Comment (jbedard) [inline nodejs/extensions.bzl:11]: Revert? 🤷

File: nodejs/extensions.bzl (hunk 1 above)
Comment (redsun82) [reply nodejs/extensions.bzl:76]: Will do so tomorrow!

File: nodejs/extensions.bzl (hunk 2 above)
Comment (redsun82) [reply nodejs/extensions.bzl:41]: it's now a `toolchain` rather than a `version`, but I can revert it.

File: nodejs/extensions.bzl (hunk 1 above)
Comment (redsun82) [reply nodejs/extensions.bzl]: Done!

File: nodejs/extensions.bzl (hunk 2 above)
Comment (redsun82) [reply nodejs/extensions.bzl]: Done!

File: nodejs/extensions.bzl (hunk 3 above)
Comment (redsun82) [reply nodejs/extensions.bzl]: Done!

**Comment by redsun82** [comment]
@jbedard thanks for the review.

While going over the code again I figured there was also a problem in the existing conflict check, in that it didn't check a difference on the `include_headers` setting. I've refactored the check so that it takes that into account and any new attributes we might add in the future. It makes the diff bigger, but I think it's worth it.

**Comment by jbedard** [comment]
Are there any tests in the WORKSPACE version that can maybe be copied for the extension?

**Comment by redsun82** [comment]
@jbedard I'll try to take a quick look.

By the way: CI failed on a temporary nodejs.org failure, which exactly the kind of problem I want to solve with this in my code base with mirrors :slightly_smiling_face: I didn't see any way I could rerun the check, but maybe I didn't look enough. On the other hand if I commit something more I'll get the reruns any way.

File: nodejs/extensions.bzl [hunk 4]
```diff
@@ -14,54 +26,65 @@ def _toolchain_extension(module_ctx):
                     # Prioritize the root-most registration of the default node toolchain version and
                     # ignore any further registrations (modules are processed breadth-first)
                     continue
-                if toolchain.node_version == registrations[toolchain.name].node_version and toolchain.node_version_from_nvmrc == registrations[toolchain.name].node_version_from_nvmrc:
+                toolchain_info = _toolchain_info(toolchain)
+                registered_toolchain_info = _toolchain_info(registrations[toolchain.name])
+                if toolchain_info != registered_toolchain_info:
```
Comment (jbedard) [inline nodejs/extensions.bzl:31]: Lets change this to `if not toolchains_equal(toolchain, registrations[toolchain.name])` just to hide most of this stuff (the `_toolchain_info` calls) in one method that describes what's going on.

**Comment by jbedard** [comment]
Sorry I forgot about this.

If @redsun82 doesn't respond maybe you can update this? I assume my latest suggestion is easy enough, and would have saved me more time trying to remember what this PR is doing that I think it's worth it 😅

**Review by jbedard — APPROVED** [review]

**Pull request merged**